lib/simplecov/source_file.rb

```diff
@@ -191,6 +337,205 @@ def process_skipped_lines(lines)
+    # Select the covered branches.
+    #
+    # @return [Array]
+    #
+    def covered_branches
+      @coverd_branches ||= root_branches.flat_map do |root_branch|
+        root_branch.sub_branches(branches).select(&:covered?)
+      end
+    end
+
+    #
+    # Select the missed branches with coverage equal to zero
+    #
+    # @return [Array]
+    #
+    def missed_branches
+      @missed_branches ||= root_branches.flat_map do |root_branch|
+        root_branch.sub_branches(branches).select(&:missed?)
+      end
+    end
+
+    #
+    # Select the perent branches inside the branches hash
+    #
+    # @return [Array]
+    #
+    def root_branches
+      @root_branches = branches.select(&:root?)
+    end
+
+    #
+    # Method check if line is branches
+    #
+    # @param [Integer] line_number
+    #
+    # @return [Boolean]
+    #
+    def branchable_line?(line_number)
+      branches_report.keys.include?(line_number)
+    end
+
+    #
+    # Return String with branches message match to the line given
+    #
+    # @param [Integer] line_number
+    #
+    # @return [String] ex: "[1, '+'],[2, '-']" two times on negative branch and non on the positive
+    #
+    def branch_per_line(line_number)
+      branches_report[line_number].each_with_object(" ".dup) do |data, message|
+        separator = message.strip.empty? ? " " : ", "
+        message << (separator + data.to_s)
+      end.strip
+    end
+
+    #
+    # Check if any branches missing on given line number
+    #
+    # @param [Integer] line_number
+    #
+    # @return [Boolean]
+    #
+    def line_with_missed_branch?(line_number)
+      return unless branchable_line?(line_number)
+      branches_report[line_number].select { |count, _sign| count.zero? }.any?
+    end
+
+    #
+    # Build full branches report
+    # Root branches represent the wrapper of all condition state that
+    # have inside the branches
+    #
+    # @return [Hash]
+    #
+    def build_branches_report
+      root_branches.each_with_object({}) do |root_branch, statistics|
+        statistics.merge!(condition_report(root_branch))
+      end
+    end
+
+    #
+    # Create hash as branches coverage report
+    # keys: lines numbers matching the branch start line
+    # Values: Array with matched branches data
+    #
+    # @param [Array] branches
+    #
+    # @return [Hash] ex: {
+    #   1 => [[1,"+"], [0, "-"]],
+    #   4 => [[10, "+"]]
+    # }
+    #
+    def condition_report(root_branch)
+      if root_branch.inline_branch?(branches)
+        inline_condition_report(root_branch)
+      else
+        multiline_condition_report(root_branch)
+      end
+    end
+
+    #
+    # Collect the information from all sub branches reports
+    #
+    # @param [Branch object] root_branch
+    #
+    # @return [Hash]
+    #
+    def multiline_condition_report(root_branch)
+      root_branch.sub_branches(branches).each_with_object({}) do |branch, cov_report|
+        cov_report[branch.start_line - 1] = [branch.report]
```

Comment:
Does this zero-base the line numbers? Is that compatible with the rest of the code?